Qemu driver: graceful shutdown feature

cheeseprocedure opened the pull request:
This PR enables the Qemu driver to attempt a graceful shutdown of guest VMs by passing an ACPI-shutdown via the [Qemu monitor](https://en.wikibooks.org/wiki/QEMU/Monitor).

This is a best-effort mechanism for notifying guests of their impending termination. The driver will still forcibly terminate the process if the kill timeout expires.

I think the integration tests can be improved; any suggestions in that area are especially appreciated.

Thank you!

schmichael commented on client/driver/qemu.go:
```diff
@@ -381,9 +431,26 @@ func (h *qemuHandle) Signal(s os.Signal) error {
+			defer monitorSocket.Close()
+			h.logger.Printf("[DEBUG] driver.qemu - sending graceful shutdown command to qemu monitor socket at %s", h.monitorPath)
+			_, err = monitorSocket.Write([]byte(qemuGracefulShutdownMsg))
+			if err == nil {
```
Is there any way to block until either the vm exits or the kill timeout is reached and make sure to always call `h.executor.ShutDown()` to kill Nomad's sidecar process?

I'm afraid this could leak VMs and Nomad executor processes.

schmichael commented on client/driver/qemu.go:
```diff
@@ -381,9 +431,26 @@ func (h *qemuHandle) Signal(s os.Signal) error {
+			if err == nil {
+				return nil
+			}
+			h.logger.Printf("[WARN] driver.qemu - failed to send '%s' to monitor socket '%s': %s", qemuGracefulShutdownMsg, h.monitorPath, err)
```
Using `%q` instead of `'%s'` will properly escape and quote strings for what it's worth. Not a big deal to change as we're not consistent either, but `%q` is pretty handy.

schmichael commented on client/driver/qemu.go:
```diff
@@ -395,13 +462,13 @@ func (h *qemuHandle) Kill() error {
 	case <-h.doneCh:
 		return nil
 	case <-time.After(h.killTimeout):
+		h.logger.Printf("[DEBUG] driver.qemu - kill timeout exceeded")
```
Can you include monitorPath and/or userPid in log messages so users have some way of connectin gthem to a particular VM?

schmichael commented on client/driver/qemu.go:
```diff
@@ -414,7 +481,7 @@ func (h *qemuHandle) run() {
 	if ps.ExitCode == 0 && werr != nil {
 		if e := killProcess(h.userPid); e != nil {
-			h.logger.Printf("[ERR] driver.qemu: error killing user process: %v", e)
+			h.logger.Printf("[ERR] driver.qemu - error killing user process: %v", e)
```
Include userPID

schmichael commented on client/driver/qemu_test.go:
```diff
@@ -14,6 +14,14 @@ import (
 	ctestutils "github.com/hashicorp/nomad/client/testutil"
 )
 
+func generateString(length int) string {
```
Replace with `strings.Repeat("x", n)`

schmichael commented on client/driver/qemu_test.go:
```diff
@@ -121,34 +134,36 @@ func TestQemuDriverUser(t *testing.T) {
+			Config: map[string]interface{}{
+				"image_path":        "linux-0.2.img",
+				"accelerator":       "tcg",
+				"graceful_shutdown": false,
```
Adding a graceful_shutdown test would be nice. Maybe copy StartOpen_Wait, but don't worry about the Open part (that's the restore path)?

schmichael replied:
(Done! Thanks!)

schmichael commented on website/source/docs/drivers/qemu.html.md:
```diff
@@ -47,6 +48,8 @@ The `qemu` driver supports the following configuration in the job spec:
   If the host machine has `qemu` installed with KVM support, users can specify
   `kvm` for the `accelerator`. Default is `tcg`.
 
+* `graceful_shutdown` `(bool: false)` - Using the [qemu monitor](https://en.wikibooks.org/wiki/QEMU/Monitor), send an ACPI shutdown signal to virtual machines rather than simply terminating them. This emulates a physical power button press, and gives instances a chance to shut down cleanly. If the VM is still running after ``kill_timeout``, it will be forcefully terminated. (Note that [prior to q ...
```
Wrap at 80 characters and indent lines with 2 spaces (we're considering loosening this style, but now we're trying to stay consistent).

> If the VM is still running after ``kill_timeout``, it will be forcefully terminated.

Huh, I thought if you wrote the shutdown message to the monitor's socket properly the `Kill()` method returned and Nomad would consider the VM dead even if it wasn't? Perhaps my comment on that code above is wrong?

cheeseprocedure commented:
Hi @schmichael - thank you for all the feedback!

I've made a number of improvements in response, including more actionable logging, fixes to ``Kill()`` to ensure we don't leak processes, and additional tests of graceful shutdown logic.

Please let me know if I can perform any additional changes and/or cleanup.

schmichael commented on client/driver/qemu_test.go:
```diff
@@ -148,7 +158,96 @@ func TestQemuDriverUser(t *testing.T) {
+	monitorPathExists := false
+	for i := 0; i < 5; i++ {
+		if _, err := os.Stat(monitorPath); !os.IsNotExist(err) {
+			fmt.Printf("Monitor socket exists at %q\n", monitorPath)
```
`d.logger.Printf` (or a `testLogger()`) is preferred but not a blocker.

schmichael commented on client/driver/qemu_test.go:
```diff
@@ -148,7 +158,96 @@ func TestQemuDriverUser(t *testing.T) {
+			monitorPathExists = true
+			break
+		}
+		time.Sleep(1 * time.Second)
```
We test on fast developer machines and very slow VMs, so testing more often (eg every 200ms) and more times (eg 50 times) would be preferred.

schmichael commented on client/driver/qemu_test.go:
```diff
@@ -148,7 +158,96 @@ func TestQemuDriverUser(t *testing.T) {
+	}
+
+	// Clean up
+	if err := resp.Handle.Kill(); err != nil {
```
Put this in a defer

schmichael commented on client/driver/qemu_test.go:
```diff
@@ -148,7 +158,96 @@ func TestQemuDriverUser(t *testing.T) {
+
+	// Clean up
+	if err := resp.Handle.Kill(); err != nil {
+		fmt.Printf("\nError killing Qemu test: %s", err)
```
Same as above (use a testLogger)

cheeseprocedure commented:
Thanks for the additional feedback! I've made some small changes in response.

schmichael commented on GNUmakefile:
```diff
@@ -154,29 +154,29 @@ lint-deps: ## Install linter dependencies
+	# 	./...
 	@echo "==> Spell checking website..."
-	@misspell -error -source=text website/source/
+	# @misspell -error -source=text website/source/
```
Ha, I don't think you meant to commit this. :)

cheeseprocedure replied:
D'oh! Fixed.

schmichael commented on client/driver/qemu_test.go:
```diff
@@ -148,7 +160,98 @@ func TestQemuDriverUser(t *testing.T) {
+	}
+
+	// Clean up
+	defer func() {
```
Oh sorry, meant to do this block in a defer right after you check the error from `d.Start`. That way you cleanup even if a fatal error is hit.

cheeseprocedure replied:
Double d'oh. Fixed!

dadgar commented on client/driver/qemu.go:
```diff
@@ -29,9 +32,15 @@ var (
+	// The key populated in Node Attributes to indicate presence of the Qemu driver
+	qemuDriverAttr                = "driver.qemu"
+	qemuDriverVersionAttr         = "driver.qemu.version"
+	qemuDriverLongMonitorPathAttr = "driver.qemu.longsocketpaths"
```
I am not a fan of this being an attribute. I think we can just check in the start method

dadgar replied:
For scheduling purposes, constraining on this can be replicated by: https://www.nomadproject.io/docs/job-specification/constraint.html#quot-version-quot-

cheeseprocedure replied:
Thanks - ``qemuDriverLongMonitorPathAttr`` is no longer an attribute!

dadgar commented on client/driver/qemu.go:
```diff
@@ -127,9 +149,26 @@ func (d *QemuDriver) Fingerprint(cfg *config.Config, node *structs.Node) (bool,
+	// Relevant fix is here:
+	// https://github.com/qemu/qemu/commit/ad9579aaa16d5b385922d49edac2c96c79bcfb6
+	currentQemuSemver := semver.New(currentQemuVersion)
+	fixedSocketPathLenVer := semver.New("2.10.1")
```
Can you make this a variable with a comment. Can be a singleton.

cheeseprocedure replied:
Done!

dadgar commented on client/driver/qemu.go:
```diff
@@ -64,6 +75,13 @@ type qemuHandle struct {
 	doneCh         chan struct{}
 }
 
+func getMonitorPath(dir string, longPathSupport string) (string, error) {
```
Comment on this method.

cheeseprocedure replied:
Done!

dadgar commented on client/driver/qemu.go:
```diff
@@ -64,6 +75,13 @@ type qemuHandle struct {
 }
 
+func getMonitorPath(dir string, longPathSupport string) (string, error) {
+	if len(dir) > legacyMaxMonitorPathLen && longPathSupport != "1" {
```
Instead of taking the `longPathSupport` string make this a function on the `QemuDriver` and use the node that is in the embedded `DriverCtx` to check the version of the qemu driver and determine if there is support.

cheeseprocedure replied:
Done!

dadgar commented on client/driver/qemu.go:
```diff
@@ -190,6 +229,19 @@ func (d *QemuDriver) Start(ctx *ExecContext, task *structs.Task) (*StartResponse
+			d.logger.Printf("[DEBUG] driver.qemu: got monitor path OK: %s", monitorPath)
+			args = append(args, "-monitor", fmt.Sprintf("unix:%s,server,nowait", monitorPath))
+		} else {
+			d.logger.Printf("[WARN] driver.qemu: %s", err)
```
I think this should fail the task.

cheeseprocedure replied:
Done! This will now return (and log) an error.

dadgar commented on client/driver/qemu.go:
```diff
@@ -381,27 +434,36 @@ func (h *qemuHandle) Signal(s os.Signal) error {
-		if h.pluginClient.Exited() {
-			return nil
+	// First, try sending a graceful shutdown command via the qemu monitor
+	err := sendQemuShutdown(h.logger, h.monitorPath, h.userPid)
```
Wrap this in a if check seeing if graceful shutdown is enabled and log the error

cheeseprocedure replied:
Done!

dadgar commented on client/driver/qemu.go:
```diff
@@ -427,3 +489,23 @@ func (h *qemuHandle) run() {
+	if monitorPath == "" {
+		logger.Printf("[DEBUG] driver.qemu: monitorPath not set; will not attempt graceful shutdown for user process pid %d", userPid)
+		return errors.New("monitorPath not set")
+	} else {
```
You can remove the else statement since the if returns.

cheeseprocedure replied:
Done!

dadgar commented on client/driver/qemu.go:
```diff
@@ -427,3 +489,23 @@ func (h *qemuHandle) run() {
+		return errors.New("monitorPath not set")
+	} else {
+		monitorSocket, err := net.Dial("unix", monitorPath)
+		if err == nil {
```
Swap the checking. `if err != nil { log && return err}` and de-indent the block.

cheeseprocedure replied:
Done!

dadgar commented on client/driver/qemu.go:
```diff
@@ -29,9 +32,15 @@ var (
+	// Reference: https://en.wikibooks.org/wiki/QEMU/Monitor
+	qemuGracefulShutdownMsg = "system_powerdown\n"
+	legacyMaxMonitorPathLen = 108
+	qemuMonitorSocketName   = "qemu-monitor.sock"
```
What about windows?

cheeseprocedure replied:
Good question! Until today, I didn't realize QEMU was available on Windows hosts. I would need to spin up an EC2 instance and experiment a bit with QEMU monitor interaction on that platform.

It appears an existing QEMU driver option can cause trouble on Windows hosts (``accelerator = "kvm"``). As a first pass, would it be acceptable to fail a task if graceful shutdown has been enabled for the job and ``runtime.GOOS == "windows"``?

dadgar replied:
Yeah I think that is acceptable. Would you mind adding error cases for both graceful and kvm on windows.

cheeseprocedure commented:
@dadgar thank you very much for your feedback! I believe I've addressed the comments, with the exception of one outstanding question about Windows support.

dadgar commented on client/driver/qemu.go:
```diff
@@ -233,13 +242,13 @@ func (d *QemuDriver) Start(ctx *ExecContext, task *structs.Task) (*StartResponse
-			d.logger.Printf("[WARN] driver.qemu: %s", err)
+		monitorPath, err := d.getMonitorPath(ctx.TaskDir.Dir)
+		if err != nil {
+			d.logger.Printf("[ERR] driver.qemu: could not get qemu monitor path - error: %s", err)
```
"[ERR] driver.qemu: could not get qemu monitor path: %s"

cheeseprocedure replied:
Done!

dadgar commented on client/driver/qemu.go:
```diff
@@ -436,11 +445,9 @@ func (h *qemuHandle) Signal(s os.Signal) error {
-	// If we did not send a graceful shutdown via the monitor socket, we'll
-	// issue an interrupt to the qemu process as a last resort
-	if err != nil {
+	if err := sendQemuShutdown(h.logger, h.monitorPath, h.userPid); err != nil {
```
This will log several debug messages relating to graceful shutdown even if the user didn't set the option. We should only attempt the graceful shutdown if the user asked for it.

cheeseprocedure replied:
Done!

cheeseprocedure commented:
@dadgar thank you for the review! In addition to the fixes above, I've added simple checks for the graceful-shutdown feature and use of the KVM accelerator which return errors if run on a Windows host.

dadgar commented:
@cheeseprocedure Awesome work! I am going to merge and add some documentation that the graceful shutdown is only supported on Linux for now!
